### PR TITLE
[Console] Add ProgressBar::preventRedrawFasterThan() and forceRedrawSlowerThan() methods

### PR DESCRIPTION
**Pull request by ostrolucky**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | TBA

The way ProgressBar redraw frequency works currently requires to know speed of progress beforehand, which is impossible to know in some situations, e.g. when showing progress of download, or I/O speed. Setting frequency too low relative to progress speed throttles I/O speed and makes progress bar flicker too much, setting it too high makes progress bar unresponsive. Current behaviour IMHO undermines usefulness of ProgressBar.

This is an attempt to replace this with more consistent experience, not requiring to know speed of progress.)

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
A redraw should be triggered every x ms, and everything y %, whichever comes first. Would be the best IMHO.

**Comment by ostrolucky** [comment]
TBH I don't see how that's useful. Anyway, that would be separate option. Neither old nor new option is based on %. So I would leave that to separate PR if someone wants that.

**Comment by nicolas-grekas** [comment]
By "%" I mean "step" of course. The display should be updated whenever the bar grows by more than Y "pixels", or has been redrawn since 1s. That's useful for DX, showing something happens.

**Comment by ostrolucky** [comment]
I think you don't understand you can't do that easily with current frequency. Current approach neither works with `%` and neither with `pixels`. If I set $max to 141977190, how would I make ProgressBar redraw itself each 1% / each X pixels with no relation to speed I'm advancing? I have created this PR to solve that issue.

> That's useful for DX, showing something happens.

Yes so set it to redraw each 0.1 seconds and be done with it. What's the use case when that's not enough?

**Comment by nicolas-grekas** [comment]
> Yes so set it to redraw each 0.1 seconds

that's not enough with fast moving progress bar, where this strategy creates visual sacades (while the current step based logic doens't) - so no, it's not enough in all cases

**Comment by ostrolucky** [comment]
I disagree that 10FPS is jerky for progress bar. But you are free to set lower frequency. Also default value in Symfony 5 can be point of discussion.

I don't see how is it useful if progress bar redraws itself more often than human eye can see. If progress bar goes from 0 to 100% in less than 100ms/10ms/1ms, do you really consider it useful to show each step all the way? There is a point when it's useless to redraw progress bar because you won't notice it changed in between time interval anyway.

**Comment by nicolas-grekas** [comment]
Yes, 100ms is way too slow for the eye. 40ms is the absolute minimum visual refresh period (24fps).
You're telling me I can set to 40ms. But that's something that I don't have to care about with the current logic, based on steps. With the time based redraw you're proposition, I suddently will have to. For this reason, your proposal is not enhancing the situation yet - it's just moving the problem somewhere else IMHO.
I really doubt it is really more complex to have code deal with both conditions to trigger a redraw.

**Comment by ostrolucky** [comment]
You won't have to care about that in Symfony 5, where there will be default FPS set. You also won't have to then care about other problems I mentioned, like calculating percentage (which is impossible to do if $max is unknown), negative performance impact and unresponsive progress bar.

That said, I don't care so much about deprecating it, it's your codebase. I consider time-based frequency superior on all fronts, but will leave old frequency in place if symfony team decides.

If you want to redraw based on whichever condition is triggered first, I would need you to consider these things first though:
- Default redraw frequency is 1/step. If I set time-based redraw frequency, it will never be applied, because default frequency will be always triggered first. User specified time-based redraw frequency, but this option will be ignored, unless he increases old frequency as well. Do you think this is ok, or how do you propose code to handle such situation?
- My problem I'm solving with this PR is that $max is unknown and speed of advancing is also unknown. Therefore I don't care about redraw frequency based on steps at all. Are you aware of the fact that in such case, if I want time-based redraw frequency always apply instead of step-based redraw frequency, I will have to call `setRedrawFrequency(PHP_INT_MAX)` for time-based redraw frequency to be usable?

Before doing any more work around this, I would also like to hear other opinions first. Thanks!

**Comment by ostrolucky** [comment]
Bumping this, feature freeze is coming.

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (rpkamp) [inline src/Symfony/Component/Console/Helper/ProgressBar.php]: This is the wrong place to update `lastWriteTime` since there is no check if `display()` has been called at all, making this entire change moot.

Also I think it would be nicer to move this entire logic to the `display` function and not mix responsibility of setting progress and deciding whether or not display the changed progress in one function. So the display function will be called every time `setProgress` is called and it can then decide if it actually wants to display something.

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Ok I will move it to display.

Rest: This is not a point of this PR. I am following current logic. You can try to do it in different PR.

Comment (rpkamp) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: I'd say either move all of to display or none of it and set lastWriteTime in both cases where display is called. Moving only lastWriteTime to display would be a very nasty case of temporal coupling.

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Ok, no moving then.

**Comment by ostrolucky** [comment]
I've got feedback people don't really understand what is this for. So I guess that's why is this not gaining any traction. I'll try to simplify explanation what is this for then:
___
First, let's ask what is redraw frequency in ProgressBar for, why would you need to adjust it? Several reasons:
1. To avoid flickering
1. To increase performance
1. To decrease amount of writes to non-console streams

Let's show example:

```php
$maxSteps           = 100000000;
$redrawFrequency    = 1000000;
$stepSize           = 100;

$progressBar = new ProgressBar(new ConsoleOutput(), $maxSteps);
$progressBar->setRedrawFrequency($redrawFrequency);
for ($i=0; $i<= $maxSteps; $i += $stepSize) {
    $progressBar->setProgress($i);
}
```


This is the output on my machine, might be different on yours:
ezgif-4-6f45fb2a40


Now let's set redrawFrequency **wrong**:
```diff
- $redrawFrequency    = 1000000;
+ $redrawFrequency    = 1;
```

Result: 
ezgif-4-ad817e1834




Now, bear in mind these are GIFs, they don't have much FPS. Normally, you will see much more flickering. But this is good enough, you see plenty flickering even there, in second GIF.

So, what can you see in these GIFs?
1. First GIF looks more smooth, despite doing less redraws
2. First progress finishes MUCH quicker. To be precise, progressbar with original settings finishes in 1,6s, second progressbar takes 146,52s, that's 90x slower!

Now, you understand importance of setting correct redraw frequency. So, what am I solving with this PR?
1. Without this patch, to set correct parameters of snippet I have shown, you absolutely need to know `$maxSteps` to be able to set correct `$redrawFrequency`. OR, you need to know how often is `setProgress` going to be called, so you can estimate redrawFrequency based on that. In real life, lot of times this is **unknown**, so you *can't* set this up, you can only estimate this. With my PR, you **don't need to care**. For real, change second constructor argument of ProgressBar in my example to `null` and play with it.
2. Problem is you need to care about `redrawFrequency` at all. My ultimate goal is to set sane default settings out of the box (this is what is 100ms hardcoded in my patch for), so you won't need to play with it at all to have reasonable redraw speed without jeopardizing performance. Working everytime, with no dependency on `$maxSteps`, no dependency on speed of your task, without having to set anything.

### Real use case:

I have a project called [stdinho](https://github.com/ostrolucky/stdinho) which uses ProgressBar to show progress of stdin/stdout streaming speed. Since input is via stdin, program has no idea about size it will get beforehand. I could estimate redraw frequency, *if I knew speed of streamin*. This speed is very different depending if I/O is via network, if server is slow, what type of disk is used etc. So it's impossible to set it correctly and setting it incorrectly have major performance penalty for speed of this streaming. So I had to give up on Symfony ProgressBar, I copied it and did changes I needed to its code directly.

**Comment by xabbuh** [comment]
If we merge this feature, can't it be implemented in a way that does not delete the old API (e.g. by creating a new class that internally uses the existing one)? I see a lot of use cases that work pretty well with the current approach and do not need to be rewritten it not entirely necessary.

**Comment by ostrolucky** [comment]
I mean, old API don't have to be deleted, but the way I see it, it just becomes redundant. Old frequency is more limited and more easy to burn yourself by setting wrong frequency. Also, name of old API is wrong for whole time, because frequency is usually dependent on time. From wiki:
> Frequency is the number of occurrences of a repeating event per unit of time.

Old API lies. It isn't dependent on time. New frequency conforms to definition.

We can also ease transition by setting default argument *now*, so all that users need to do in 4.x is just remove setRedrawFrequency call.

ProgressBar is very human-facing part and humans don't care so much if old and new frequency is exactly same, so this isn't covered via BC promise.

We can also not remove setRedrawFrequency, but add to it second, boolean argument which signals time-based frequency. Deprecation will complain if it isn't set, but will stop complaining in Symfony 5 and just consider input to be time-based frequency. This means applications will keep working with no changes. Downside is if given argument isn't adjusted, ProgressBar will be redrawn much less frequently than intended originally, so not sure if this is a way.

We can also play a long game and just deprecate calling setRedrawFrequency with float value. And in Symfony 5 it will be allowed again, but will mean this is time-based frequency.

There are lot of options how to handle this, someone just needs to decide what to do.

**Comment by ostrolucky** [comment]
bump

**Comment by stof** [comment]
@ostrolucky the main focus of the core team these days is on finishing 4.1, as the release is due in approximately 1 week. We'll get back to focus to all next PRs after it

**Comment by ostrolucky** [comment]
Would approach of introducing new class `OutputThrottledByFrequency` which is just a decorator of OutputInterface be more welcome?

New diff of ProgressBar would be following

```diff
@@ -31,7 +32,7 @@ final class ProgressBar
     private $progressChar = '>';
     private $format;
     private $internalFormat;
-    private $redrawFreq = 1;
+    private $redrawFreq;
     private $output;
     private $step = 0;
     private $max;
@@ -57,6 +58,10 @@ final class ProgressBar
             $output = $output->getErrorOutput();
         }

+        if (!$output instanceof OutputThrottledByFrequency) {
+            $output = new OutputThrottledByFrequency($output, .1);
+        }
+
         $this->output = $output;
         $this->setMaxSteps($max);
         $this->terminal = new Terminal();
@@ -64,9 +69,6 @@ final class ProgressBar
         if (!$this->output->isDecorated()) {
             // disable overwrite when output does not support ANSI codes.
             $this->overwrite = false;
-
-            // set a reasonable redraw frequency so output isn't flooded
-            $this->setRedrawFrequency($max / 10);
         }

         $this->startTime = time();
@@ -237,10 +239,15 @@ final class ProgressBar
      * Sets the redraw frequency.
      *
      * @param int|float $freq The frequency in steps
+     *
+     * @deprecated since version 4.1, to be removed in 5.0. Use $redrawFrequency argument in constructor instead.
      */
     public function setRedrawFrequency(int $freq)
     {
+        @trigger_error(sprintf('The %s() method is deprecated since Symfony 4.1 and will be removed in 5.0. Use $redrawFrequency argument in constructor instead.', __METHOD__), E_USER_DEPRECATED);
+
         $this->redrawFreq = max($freq, 1);
+        $this->output->setRedrawFrequency(0);
     }

     /**
@@ -291,7 +298,7 @@ final class ProgressBar
         $currPeriod = (int) ($step / $this->redrawFreq);
         $this->step = $step;
         $this->percent = $this->max ? (float) $this->step / $this->max : 0;
-        if ($prevPeriod !== $currPeriod || $this->max === $step) {
+        if ($this->redrawFreq === null || $prevPeriod !== $currPeriod || $this->max === $step) {
             $this->display();
         }
     }
```

And decorator is simple
```php
final class OutputThrottledByFrequency implements OutputInterface
{
    private $output;
    private $redrawFrequency;
    private $lastWriteTime;

    /**
     * @param float $redrawFrequency Frequency of redrawing in seconds
     */
    public function __construct(OutputInterface $output, float $redrawFrequency)
    {
        $this->output = $output;
        $this->redrawFrequency = $redrawFrequency;
    }

    public function setRedrawFrequency(float $redrawFrequency): void
    {
        $this->redrawFrequency = $redrawFrequency;
    }

    /**
     * {@inheritdoc}
     */
    public function write($messages, $newline = false, $options = 0): void
    {
        $time = microtime(true);

        if ($time < $this->lastWriteTime) {
            return;
        }

        $this->output->write($messages, $newline, $options);
        $this->lastWriteTime = $time;
    }
...
}
```

**Comment by ostrolucky** [comment]
Update:
- rebased
- made tests pass
- bumped FPS, no longer configurable though (see screencast, I think it should be enough for everybody now)
- made this default (this isn't API break so why bother with ridiculously complicated upgrade strategy)
- dropped deprecation (since I see it's controversial, might be better to be done in separate PR if anybody decides to do that)

Some visualization:
- setRedrawFrequency(1) (which was default before this PR): https://asciinema.org/a/odDzjluH2BhJL3xjsc1qkFZ1m (cut short because it was over 2 minutes and couldn't upload)
- 10FPS https://asciinema.org/a/QMFGYG4AuVW8HTfj7SamkjGdS
- 20FPS https://asciinema.org/a/38bzKbeEdnQT5uSu2T9HQA874

@nicolas-grekas is this the jerkyness you talked about you wanted to solve? https://asciinema.org/a/EyZ89dHtAxmrSquB5hklfd2Bd I found it can't be solved when combined with both, because time & step % both race between each other

**Comment by ostrolucky** [comment]
bump

**Comment by ostrolucky** [comment]
bump

**Comment by apfelbox** [comment]
I had several issues that triggering to many draws in the progress bar actually slowed down my process by a factor of 4-5. So I would love this change as it is good enough ™️ and solves several real issues.
And not sure whether my CLI actually needs 60 FPS... 20 should be enough and we could even increase it in the future.

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Correct, this changes default behaviour.

Correct, it's not possible to revert back to time-based redraw once progressbar object has been constructed and setRedrawFrequency called.

I don't think this change will cause flooding the output of any existing app,  because that happens only if app produces lot of output. If it produces lot of output and it's not flooding the output, that means app developers already use setRedrawFrequency call, which will mean new default behaviour won't apply.

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: >  it's not flooding the output, that means app developers already use setRedrawFrequency call

Why so? the default right now is 1/10, so no call to setRedrawFrequency is required to prevent flooding. Did I miss something?

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: 1/10 default does not prevent flooding at all. If there is `2*10^6` redraws per second, reducing it to `2*10^5` does not help much. It was really arbitrarily set frequency which helps in minimum cases. Additionally, new default will help even in such cases where 1/10 is correct, since 50ms is correct as well, in all cases.

I'm trying to think about use case with frequency where redrawing every 10th time is more correct than redrawing every 50ms, but I can't find any :thinking: Can you think about some?

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: the frequency is not in time but in steps: 1/10 means there will always be 10 steps whatever the $max.

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: It's neither. 1/10 means redraw will happen every $max/10 time. So it triggers 10 redraws in total _only_ if $max is 100. If $max is 1000 and job takes 10 seconds to finish, there are 100 redraw calls each second with current default. It's really pretty arbitrary number and not very useful default. It might help only if by coincidence current `time of the job run/$max/10` approaches maximum writing speed of output (terminal).

Here's minimal reproducible so you can test it:
```php
<?php
require 'vendor/autoload.php';

use Symfony\Component\Console\Helper\ProgressBar;
use Symfony\Component\Console\Output\ConsoleOutput;

$progressBar = new ProgressBar(new ConsoleOutput(), $maxSteps = 100000000);
for ($i=0; $i<= $maxSteps; $i++) {
    $progressBar->advance();
}
```

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Actually you are right. However, default you are talking about is applied only when output is not decorated. I think limit to 10 redraws as default makes sense there. I've put it back.

For non-decorated output, default is that it's being redrawn on each step and I would like to change that to something more reasonable (this PR).

File: src/Symfony/Component/Console/Tests/Helper/ProgressBarTest.php
Comment (ostrolucky) [inline src/Symfony/Component/Console/Tests/Helper/ProgressBarTest.php]: I've changed this here because there is racing condition in this test when going with usleep approach. Sometimes, last output is `< 3 sec` instead of `< 1 sec`.

**Comment by ostrolucky** [comment]
What needs to be done here to move this forward?

**Comment by Simperfit** [comment]
cc @fabpot this one seems useful and it seems ready IMHO.

**Review by Simperfit — APPROVED** [review]

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (javiereguiluz) [inline src/Symfony/Component/Console/Helper/ProgressBar.php]: We could change this to: `intdiv($prevStep, $this->redrawFreq)`

File: src/Symfony/Component/Console/Tests/Helper/ProgressBarTest.php
Comment (javiereguiluz) [inline src/Symfony/Component/Console/Tests/Helper/ProgressBarTest.php]: Variables whose names describe their contents are not future proof :)  If we change this in the future to `40001`, the code will be confusing. We can rename this to `REDRAW_INTERVAL`, `REDRAW_TIME`, `FRAME_DURATION` or anything else we like. Thanks.

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (javiereguiluz) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Humans are used to 24, 30 and 60 frames per second ... which means a frame duration of 41ms, 33ms and 16ms. So, maybe we can change `50ms` (which seems arbitrary) to `41ms`? (Maybe 50ms is not arbitrary and there's a reason behind it).

**Review by javiereguiluz — COMMENTED** [review]
@ostrolucky I like this ... but there's the usual "BC policy" applied to Symfony for changes like this. I think we need to find a satisfactory solution.

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Redraw every 50ms corresponds to 20FPS. This is best FPS that can be achieved without hurting performance too much. Even this is too much IMO and it has negative performance impact. Remember, this is CLI application, not a movie. Even GIFs have 15FPS and they are perfectly ok - and that is for entertainment value. ProgressBar is just an indicator. If anything, frequency should be decreased. I have increased to 50ms it after complaints, but trying to go for movie like FPS in CLI is not realistic.

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Sure, even though I tried to preserve current way of doing things

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: On second thought, let's avoid doing unnecessary extra function calls

**Comment by ostrolucky** [comment]
Comment addressed.

FPS increased to 25 (redraw each 0.04 sec).

> there's the usual "BC policy"

I'm sorry, I don't see how is this affecting BC policy. It's fully backwards compatible. BC policy is for interfaces, I didn't change any. Only thing that is changed is there will be less redraws in interactive mode by default. That shouldn't be risky change.

To remind importance of this PR, snippet I posted in https://github.com/symfony/symfony/pull/26339#issuecomment-377784978 takes 142s without this PR and 2,29s with this PR (did a new benchmark).

**Comment by nicolas-grekas** [comment]
As discussed at #eufossa hackathon, we could have this spec/new methods:
```
$progress->setMaxSteps(100000); <- when not redraw freq is defined, default to max/20
$progress->rejectRedrawFasterThan(0.01); <- defaults to 0s - ie always accept
$progress->forceRedrawSlowerThan(1); <- defaults to 1s
```

**Comment by ostrolucky** [comment]
Unfortunately I can't use these defaults combined :( 

Max/20 + force redraw 1 second would make progressbar draw either once per second or once per 20%, which is too low from what people are used. Trying to figure out way to move forward.

**Comment by nicolas-grekas** [comment]
Don't hesitate to push what you have, I'd like to understand what you mean.
I you have a small repo to run the progress bar, I'm interested too :)

**Comment by ostrolucky** [comment]
Well it's simple. What I understond as suggested default step freq would cause to redraw like this by default

Untitled

Playing with basic example in https://github.com/symfony/symfony/pull/26339#issuecomment-464388538 is enough, I don't have any fancy reproducers, just play with `setRedrawFrequency` values.

Since this is hard to merge as it is, I'm for now removing change of default behaviour. Just new methods added which aid anybody experiencing issues like I have.

**Review by chalasr — APPROVED** [review]
Works for me as-is, changing the default can be discussed later on.

**Comment by nicolas-grekas** [comment]
What about this added patch? It work well with your example:
```diff
--- a/src/Symfony/Component/Console/Helper/ProgressBar.php
+++ b/src/Symfony/Component/Console/Helper/ProgressBar.php
@@ -31,7 +31,7 @@ final class ProgressBar
     private $progressChar = '>';
     private $format;
     private $internalFormat;
-    private $redrawFreq = 1;
+    private $redrawFreq;
     private $lastWriteTime;
     private $minSecondsBetweenRedraws = 0;
     private $maxSecondsBetweenRedraws = 1;
@@ -318,8 +318,9 @@ final class ProgressBar
             $step = 0;
         }
 
-        $prevPeriod = (int) ($this->step / $this->redrawFreq);
-        $currPeriod = (int) ($step / $this->redrawFreq);
+        $redrawFreq = $this->redrawFreq ?? (($this->max ?: 20) / 20);
+        $prevPeriod = (int) ($this->step / $redrawFreq);
+        $currPeriod = (int) ($step / $redrawFreq);
         $this->step = $step;
         $this->percent = $this->max ? (float) $this->step / $this->max : 0;
         $timeInterval = microtime(true) - $this->lastWriteTime;
```

**Comment by ostrolucky** [comment]
That causes same result as I posted in latest gif. And this is how it looks like without setting maximum:
ezgif-3-f4d12f9b2fa3

And of course, since this changes default, some test cases would have to be changed.

**Comment by nicolas-grekas** [comment]
@ostrolucky I push forced on your branch to rebase on latest 4.4, add a CHANGELOG, and fix what I understood from your last GIF. Am I right that the issue you wanted to showcase was the cursor moving back when preventRedrawFasterThan is used? If yes, that's fixed. If not, please share the exact code you ran because right now this is cryptic to me.

To me, this PR is ready:
- we can set a minimum time between redraws to prevent flooding the console. This is disabled by default.
- When a max step is set, we don't draw more than each 5% or each time a max duration between redraws is reached - set to 1s by default
- When no max step is set, we display an always progress bar that moves always forward, but never more often than the above minimum time between redraws

File: src/Symfony/Component/Console/Helper/ProgressBar.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Helper/ProgressBar.php]: The best default is not zero, but this would be behavioral change and would be caught by the tests
Since this is not technically a BC break, but more a behavioral change, we can maybe still change the default to e.g. 0.1?

Comment (ostrolucky) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: I would be happy to change it now and I did that, but on hackaton I was told I am changing too much testcases so I abandonded this attempt

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: Sure sorry, my question was for more for @symfony/deciders to consider when reviewing :)

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ostrolucky.

**Pull request merged**
